lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp

```diff
@@ -0,0 +1,604 @@
+
+#include <vector>
+
+#define DEBUG_TYPE "firrtl-lower-open-aggs"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+
+/// Information on non-hw (ref) elements.
+struct NonHWField {
+  /// Type of the field, not a hardware type.
+  FIRRTLType type;
+  /// FieldID relative to base of converted type.
+  uint64_t fieldID;
+  /// Relative orientation.  False means aligned.
+  bool isFlip;
+  /// String suffix naming this field.
+  SmallString<16> suffix;
+  /// Print this structure to llvm::errs().
+  void dump() const;
+};
+
+/// Mapped port info
+struct PortMappingInfo {
+  /// Preserve this port, map use of old directly to new.
+  bool identity;
+
+  // When not identity, the port will be split:
+
+  /// Type of the hardware-only portion.  May be null, indicating all non-hw.
+  Type hwType;
+  /// List of the individual non-hw fields to be split out.
+  SmallVector<NonHWField, 0> fields;
+
+  /// Determine number of types this argument maps to.
+  size_t count(bool includeErased = false) const {
+    if (identity)
+      return 1;
+    return fields.size() + (hwType ? 1 : 0) + (includeErased ? 1 : 0);
+  }
+  void dump() const;
+};
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const NonHWField &field) {
+  return os << llvm::formatv(
+             "non-HW(type={0}, fieldID={1}, isFlip={2}, suffix={3})",
+             field.type, field.fieldID, field.isFlip, field.suffix);
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+                              const PortMappingInfo &pmi) {
+  if (pmi.identity)
+    return os << "(identity)";
+
+  os << "[[hw portion: ";
+  if (pmi.hwType)
+    os << pmi.hwType;
+  else
+    os << "(none)";
+  os << ", fields: ";
+  llvm::interleaveComma(pmi.fields, os);
+  return os << "]]";
+}
+
+} // namespace
+
+void NonHWField::dump() const { llvm::errs() << *this; }
+void PortMappingInfo::dump() const { llvm::errs() << *this; }
+
+template <typename Range>
+LogicalResult walkPortMappings(
+    Range &&range, bool includeErased,
+    llvm::function_ref<LogicalResult(size_t, PortMappingInfo &, size_t)>
+        callback) {
+  size_t count = 0;
+  for (const auto &[index, pmi] : llvm::enumerate(range)) {
+    if (failed(callback(index, pmi, count)))
+      return failure();
+    count += pmi.count(includeErased);
+  }
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Visitor
+//===----------------------------------------------------------------------===//
+
+namespace {
+class Visitor : public FIRRTLVisitor<Visitor, LogicalResult> {
+public:
+  explicit Visitor(MLIRContext *context) : context(context){};
+
+  /// Entrypoint.
+  LogicalResult visit(FModuleLike mod);
+
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitDecl;
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<Visitor, LogicalResult>::visitStmt;
+
+  LogicalResult visitDecl(InstanceOp op);
+
```

Comment:
Can you have an open-typed wire? If so, does that need to be handled here?

Reply:
No, can't put refs (or aggs w/refs) in wires -- "not yet".  Ports only, arguably that's already a mismatch.  But yes I think that could be handled here, and in similar way-- split out the ref-y things and let wire be wire however it's currently used.  There are asks/uses for such a thing.  However, I'd like to get this working for what's in the spec first before we add more to it.

But totally I think that's something we can do, I think can be handled similarly here to the opensub* ops, and would also address the sort of thing that motivated `ref.pipe` re:place for refs that isn't a port as that can be inconvenient.

However before running with that, it would be good to work through... the details re:mismatches in the abstractions/concepts backing our IR and types so we have a solid foundation moving forward (particularly with fancier non-hardware things beyond refs).  For example, it'd be good to have a proper reason for why a ref is suitable for a wire but not a register.  There are various framings/ways to make this coherent, offhand I suppose I'd say it's because wires aren't actually declarations of hardware, they're named dataflow points for plumbing together actual hardware.  Maybe? :smile: .

Anyway good question, thanks!

Reply:
Yeah, I don't necessarily think wires _should_ support refs. But apparently they currently unintentionally do, this doesn't give an error (well, at least until lowering to HW):
```mlir
firrtl.extmodule @A(out ref: !firrtl.probe<uint<1>>)
firrtl.module @Test() {
  %r = firrtl.instance a @A(out ref: !firrtl.probe<uint<1>>)
  %w = firrtl.wire : !firrtl.probe<uint<1>>
  firrtl.ref.define %w, %r : !firrtl.probe<uint<1>>
}
```

Reply:
Though that's orthogonal to this pr

Reply:
Oh, indeed.  Added bits to reject this in the FIRRTL parser for now in new PR, but wire is "open" so will leave that alone at the IR level.